### PR TITLE
Support Resource Bindings in UploadWorker

### PR DESCRIPTION
**Pull request by grobinson-cloudflare**
This PR makes it possible to bind Worker scripts to KV namespaces when uploading scripts via the API.

## Types of changes

What sort of change does your code introduce/modify?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

[1]: https://help.github.com/articles/closing-issues-using-keywords/

### CODE REVIEW

File: workers.go [hunk 1]
```diff
@@ -167,15 +182,41 @@ func (api *API) ListWorkerScripts() (WorkerListResponse, error) {
 // UploadWorker push raw script content for your worker
 //
 // API reference: https://api.cloudflare.com/#worker-script-upload-worker
-func (api *API) UploadWorker(requestParams *WorkerRequestParams, data string) (WorkerScriptResponse, error) {
+func (api *API) UploadWorker(requestParams *WorkerRequestParams, data string, metadataParams *WorkerMetadataParams) (WorkerScriptResponse, error) {
```
Comment (sevki) [inline workers.go]: there is no requirement to upload metadata and changing the public facing interface would require people to update their code to pass in a second nil parameter. 

What do you think about updating this PR to keep this function as is and adding a new `UploadWorkerWithResourceBindings` function to so we don't change API?

File: workers.go [hunk 2]
```diff
@@ -167,15 +182,41 @@ func (api *API) ListWorkerScripts() (WorkerListResponse, error) {
...
+
+func (api *API) uploadWorker(requestParams *WorkerRequestParams, data string, metadataParams *WorkerMetadataParams) (WorkerScriptResponse, error) {
 	uri := "/zones/" + requestParams.ZoneID + "/workers/script"
+	if metadataParams != nil {
+		return api.doMultipartUploadWorker(uri, data, *metadataParams)
+	}
+	return api.doUploadWorker(uri, data)
+}
+
+// UploadNamedWorker push raw script content for your worker
```
Comment (sevki) [inline workers.go]: Is this referred to as a `NamedWorker` anywhere else?
@deuill?

File: workers.go [hunk 3]
```diff
@@ -186,19 +227,31 @@ func (api *API) UploadWorker(requestParams *WorkerRequestParams, data string) (W
...
+	if err != nil {
+		return r, errors.Wrap(err, errMakeRequestError)
 	}
-	uri := "/accounts/" + api.OrganizationID + "/workers/scripts/" + scriptName
 	headers := make(http.Header)
-	headers.Set("Content-Type", "application/javascript")
-	res, err := api.makeRequestWithHeaders("PUT", uri, []byte(data), headers)
-	var r WorkerScriptResponse
+	headers.Set("Content-Type", writer.FormDataContentType())
+	res, err := api.makeRequestWithHeaders("PUT", uri, body.Bytes(), headers)
```
Comment (sevki) [inline workers.go]: should we consider switching this to [`http.MethodPut`](https://golang.org/pkg/net/http/#pkg-constants) too to make it more consistent with line 219?

File: workers.go [hunk 4]
```diff
@@ -186,19 +227,31 @@ func (api *API) UploadWorker(requestParams *WorkerRequestParams, data string) (W
...
-//
-// API reference: https://api.cloudflare.com/#worker-script-upload-worker
-func (api *API) uploadWorkerWithName(scriptName string, data string) (WorkerScriptResponse, error) {
-	if api.OrganizationID == "" {
-		return WorkerScriptResponse{}, errors.New("organization ID required for enterprise only request")
+func (api *API) doMultipartUploadWorker(uri, data string, metadata WorkerMetadataParams) (WorkerScriptResponse, error) {
+	var r WorkerScriptResponse
+	body := &bytes.Buffer{}
+	writer := multipart.NewWriter(body)
+	metadataPart, err := writer.CreateFormFile("metadata", "metadata.json")
```
Comment (sevki) [inline workers.go]: we should probably not hardcode these as  `metadata.json` and `script.js` are not the only available options 
https://developers.cloudflare.com/workers/api/resource-bindings/webassembly-modules/

**Comment by patryk** [comment]
Merged a similar solution in #363. Thanks @grobinson-cloudflare !

**Pull request closed without merging**